Adds teleop support for Unitree G1 with Inspire 5-finger hand, take PickPlace task as an example

yami007007 opened the pull request:
# Description
Teleop support for Unitree G1 with Inspire 5-finger hand, take PickPlace task as an example.

# Type of change
- New feature (non-breaking change which adds functionality)

## Changes:
-Preparation of the humanoid robot asset: Unitree G1 (29 DOF) with Inspire 5-finger hand
-Configuration of the robot’s articulation
-Retargeting setup for the humanoid robot
-Creation of the manipulation task environment: pickplace_unitree_g1_inspire_hand_env_cfg.py
-Updated the how-to tutorial of [cloudxr_teleoperation.rst](https://github.com/isaac-sim/IsaacLab/pull/3242/files#diff-3896784d4dbd3e2106adae8f4ab65530a3bb87430cf8d15463aff37f3a975fec).

Below is a diagram showing the robot asset and the workflow in teleop.
![teleop_diagram](https://github.com/user-attachments/assets/4c247c6d-57d6-4d18-90ed-1b2af7827437)



# Test Steps
1. Run the teleop task with:   
python scripts/tools/record_demos.py \
    --task Isaac-PickPlace-G1-InspireFTP-Abs-v0 \
    --teleop_device handtracking \
    --enable_pinocchio    

2. Replay the collected demos with:    
python scripts/tools/replay_demos.py \
    --task Isaac-PickPlace-G1-InspireFTP-Abs-v0 \
    --enable_pinocchio \
    --dataset_file datasets/dataset.hdf5

## Screenshots
Please attach before and after screenshots of the change if applicable.
![changelog](https://github.com/user-attachments/assets/e8781d07-b327-4b56-b528-fcbafbaf15b9)

## Checklist

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `./isaaclab.sh --format`
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have updated the changelog and the corresponding version in the extension's `config/extension.toml` file
- [ ] I have added my name to the `CONTRIBUTORS.md` or my name already exists there

kellyguo11 commented on source/isaaclab_assets/isaaclab_assets/robots/unitree.py:
```diff
@@ -381,3 +381,161 @@
+"""
+G1_INSPIRE_FTP_CFG = ArticulationCfg(
+    spawn=sim_utils.UsdFileCfg(
+        usd_path=f"./usecase/humanoid_teleop/g1_29dof_rev_1_0_with_inspire_hand_FTP/g1_29dof_rev_1_0_with_inspire_hand_retarget_inspire_white.usd",
```
we should get this USD uploaded to the Isaac Sim nucleus folder so that we can make it available publicly on the S3 bucket

kellyguo11 replied:
how different is this asset from the existing G1 on S3?

yami007007 commented on source/isaaclab/isaaclab/devices/openxr/retargeters/humanoid/unitree/unitree_g1_dex_retargeting_utils.py:
```diff
@@ -0,0 +1,274 @@
+#
+# SPDX-License-Identifier: BSD-3-Clause
+
+# Copyright (c) 2022-2025, The Isaac Lab Project Developers.
```
Sure, will update in the next commit along with the correct USD asset location.

ooctipus commented on source/isaaclab/isaaclab/devices/openxr/retargeters/__init__.py:
```diff
@@ -5,6 +5,7 @@
 """Retargeters for mapping input device data to robot commands."""
 
 from .humanoid.fourier.gr1t2_retargeter import GR1T2Retargeter, GR1T2RetargeterCfg
+from .humanoid.unitree.unitree_g1_retargeter import UnitreeG1Retargeter, UnitreeG1RetargeterCfg
```
high level comment:
     robot - worflow specific definition maybe more appropriate in isaaclab_tasks rather than in core. Things in core should be mostly just pure interface how robot - worflow specific definition should use them

kellyguo11 commented on source/isaaclab/isaaclab/devices/openxr/retargeters/humanoid/unitree/unitree_g1_dex_retargeting_utils.py:
```diff
@@ -0,0 +1,259 @@
+        hand_joint_names: list[str],
+        right_hand_config_filename: str = "unitree_hand_right_dexpilot.yml",
+        left_hand_config_filename: str = "unitree_hand_left_dexpilot.yml",
+        left_hand_urdf_path: str = "omniverse://isaac-dev.ov.nvidia.com/Projects/agile/Robots/Collected_g1_29dof/g1_29dof_rev_1_0_retarget_left_hand.urdf",
```
assets will need to be uploaded to S3.

yami007007 replied:
Yes, @Sheikh is handling this.

kellyguo11 commented on source/isaaclab_assets/isaaclab_assets/robots/unitree.py:
```diff
@@ -381,3 +382,163 @@
+"""
+G1_INSPIRE_FTP_CFG = ArticulationCfg(
+    spawn=sim_utils.UsdFileCfg(
+        usd_path="omniverse://isaac-dev.ov.nvidia.com/Projects/agile/Robots/Collected_g1_29dof/g1_29dof_rev_1_0_with_inspire_hand.usd",
```
asset will need to be uploaded to S3

kellyguo11 replied:
do we have the license for this file as well?

kellyguo11 commented on source/isaaclab_tasks/isaaclab_tasks/manager_based/manipulation/pick_place/README_teleop_unitreeG1.md:
```diff
@@ -0,0 +1,147 @@
+# Teleoperation
```
can we move this to the official docs instead of having it separately in the code? imitation learning and teleoperation docs are under https://isaac-sim.github.io/IsaacLab/main/source/overview/imitation-learning/index.html. we can also make this into a how-to guide here - https://isaac-sim.github.io/IsaacLab/main/source/how-to/index.html

kellyguo11 commented on source/isaaclab_tasks/isaaclab_tasks/manager_based/manipulation/pick_place/__init__.py:
```diff
@@ -45,10 +45,10 @@
 
 gym.register(
-    id="Isaac-PickPlace-GR1T2-WaistEnabled-Abs-v0",
+    id="Isaac-PickPlace-G1-InspireFTP-Abs-v0",
```
is this intended to replace the existing environment? @michaellin6 for viz.

kellyguo11 commented on source/isaaclab_tasks/isaaclab_tasks/manager_based/manipulation/pick_place/mdp/observations.py:
```diff
@@ -24,8 +24,12 @@ def object_obs(
     body_pos_w = env.scene["robot"].data.body_pos_w
-    left_eef_idx = env.scene["robot"].data.body_names.index("left_hand_roll_link")
-    right_eef_idx = env.scene["robot"].data.body_names.index("right_hand_roll_link")
+    if env.unwrapped.cfg.robot_name == "Unitree_G1":
```
would be good to avoid the ifs by robot name. can we make the body name a parameter instead that can be define through the config?

kellyguo11 commented on source/isaaclab_tasks/isaaclab_tasks/manager_based/manipulation/pick_place/mdp/observations.py:
```diff
@@ -89,8 +105,10 @@ def get_right_eef_quat(
 ) -> torch.Tensor:
-    hand_joint_states = env.scene["robot"].data.joint_pos[:, -22:]  # Hand joints are last 22 entries of joint state
-
+    if env.unwrapped.cfg.robot_name == "Unitree_G1":
```
would be good to avoid the if statement here as well

yami007007 replied:
Updated: use a parameter instead of if statements.

kellyguo11 commented on source/isaaclab_tasks/isaaclab_tasks/manager_based/manipulation/pick_place/mdp/terminations.py:
```diff
@@ -65,7 +65,12 @@ def task_done_pick_place(
     # Get right wrist position relative to environment origin
     robot_body_pos_w = env.scene["robot"].data.body_pos_w
-    right_eef_idx = env.scene["robot"].data.body_names.index("right_hand_roll_link")
+    if env.unwrapped.cfg.robot_name == "Unitree_G1":
```
same comment as above

kellyguo11 reviewed (commented):
The gifs are quite large, can we ask Sheikh to upload them to the server similar to the other mimic gifs? https://nvidia.slack.com/archives/C07C06BV4VB/p1747662149763689. image files should be .jpg for smaller image size.

yami007007 commented:
> The gifs are quite large, can we ask Sheikh to upload them to the server similar to the other mimic gifs? https://nvidia.slack.com/archives/C07C06BV4VB/p1747662149763689. image files should be .jpg for smaller image size.

Deleted the gifs, moved the ReadMe to [docs/source/how-to/cloudxr_teleoperation.rst](https://github.com/isaac-sim/IsaacLab/pull/3242/files#diff-3896784d4dbd3e2106adae8f4ab65530a3bb87430cf8d15463aff37f3a975fec).

rwiltz commented on source/isaaclab/isaaclab/devices/openxr/retargeters/humanoid/unitree/data/configs/dex-retargeting/unitree_hand_left_dexpilot.yml:
Move source/isaaclab/isaaclab/devices/openxr/retargeters/humanoid/unitree/data/configs/dex-retargeting/unitree_hand_left_dexpilot.yml -> source/isaaclab/isaaclab/devices/openxr/retargeters/humanoid/unitree/inspire/data/configs/dex-retargeting/unitree_hand_left_dexpilot.yml

yami007007 replied:
Updated.

yami007007 commented on source/isaaclab/isaaclab/devices/openxr/retargeters/humanoid/unitree/g1_dex_retargeting_utils.py:
Updated.

rwiltz reviewed (approved):
LGTM from teleop side
